File: EPI/19.2/swift/19.2_test.swift
import Darwin

func == <T: Equatable>(lhs: [[T]], rhs: [[T]]) -> Bool {
    guard lhs.count == rhs.count else {
        return false
    }
    for (i, array) in lhs.enumerated() {
        guard array == rhs[i] else {
            return false
        }
    }
    return true
}

let tests: [(image: [[Int]], cell: Coord, flipped: [[Int]])] = [
    (
        image: [
            [0, 0, 0, 0],
            [0, 0, 0, 0],
            [0, 0, 0, 0],
            [0, 0, 0, 0],
        ],
        cell: (row: 2, col: 1),
        flipped: [
            [1, 1, 1, 1],
            [1, 1, 1, 1],
            [1, 1, 1, 1],
            [1, 1, 1, 1],
        ]
    ),
    (
        image: [
            [1, 0, 0, 0],
            [0, 1, 1, 0],
            [0, 0, 1, 0],
            [0, 0, 1, 0],
        ],
        cell: (row: 2, col: 1),
        flipped: [
            [1, 0, 0, 0],
            [1, 1, 1, 0],
            [1, 1, 1, 0],
            [1, 1, 1, 0],
        ]
    ),
    (
        image: [
            [1, 0, 0, 0],
            [0, 1, 1, 0],
            [0, 0, 1, 0],
            [0, 0, 1, 0],
        ],
        cell: (row: 1, col: 1),
        flipped: [
            [1, 0, 0, 0],
            [0, 0, 0, 0],
            [0, 0, 0, 0],
            [0, 0, 0, 0],
        ]
    ),
]

private func intToColor(_ image: [[Int]]) -> [[Color]] {
    return image.map { $0.map { Color(rawValue: $0)! } }
}

private func colorToInt(_ image: [[Color]]) -> [[Int]] {
    return image.map { $0.map { $0.rawValue } }
}

for test in tests {
    var image = intToColor(test.image)
    flipRegion(test.cell, in: &image)
    let actual = colorToInt(image)

    guard actual == test.flipped else {
        print("For image \(test.image) and cell \(test.cell), " +
            "expected flipped image to be \(test.flipped), but was \(actual)")
        exit(1)
    }
}

File: EPI/19.2/typescript/19.2_test.ts
import {Cell, flip} from './19.2';

declare function require(name: string): any;
const assert = require('assert');
const inspect = (v: any) => require('util').inspect(v, {depth: null});

const tests: {grid: boolean[][], cell: Cell, flipped: boolean[][]}[] = [
    {
        grid: [
            [true,  true,  true,  true],
            [true,  true,  true,  true],
            [true,  true,  true,  true],
            [true,  true,  true,  true],
        ],
        cell: {row: 2, col: 1},
        flipped: [
            [false, false, false, false],
            [false, false, false, false],
            [false, false, false, false],
            [false, false, false, false],
        ]
    },
    {
        grid: [
            [false, true,  true,  true],
            [true,  false, false, true],
            [true,  true,  false, true],
            [true,  true,  false, true],
        ],
        cell: {row: 2, col: 1},
        flipped: [
            [false, true,  true,  true],
            [false, false, false, true],
            [false, false, false, true],
            [false, false, false, true],
        ]
    },
    {
        grid: [
            [false, true,  true,  true],
            [true,  false, false, true],
            [true,  true,  false, true],
            [true,  true,  false, true],
        ],
        cell: {row: 1, col: 1},
        flipped: [
            [false, true,  true,  true],
            [true,  true,  true,  true],
            [true,  true,  true,  true],
            [true,  true,  true,  true],
        ]
    },
];

for (const test of tests) {
    const actual = test.grid.map(row => row.slice());
    flip(actual, test.cell);
    assert.deepStrictEqual(actual, test.flipped,
        `For grid ${inspect(test.grid)} and cell ${inspect(test.cell)}, ` +
        `expected flipped to be ${inspect(test.flipped)}, but was ${inspect(actual)}`);
}
